### PR TITLE
Fix typed array export

### LINKED ISSUE #62916: Can't assign nodes to exported array
___

***Bugsquad note:** This issue has been confirmed several times already. No need to confirm it further.*

___

### Godot version

Godot 4.0 alpha 11

### System information

Solus OS and Windows

### Issue description

Can't assign nodes to an exported array of node pointers

https://user-images.githubusercontent.com/61121689/178321358-fe05eec9-17ab-4d94-ac87-f87b1e059b97.mp4

### Steps to reproduce

1) Export an array of node pointers (`@export var node_pointer_array: Array[Node2D]`)
2) Try to assign an element to the array on the editor
3) Element will not be assigned

Or: check out the minimal reproduction project below

### Minimal reproduction project

[export array of node pointers.zip](https://github.com/godotengine/godot/files/9086249/export.array.of.node.pointers.zip)

### PR DESCRIPTION
**Pull request by GuilhermeGSousa**
Fixes #62916. 

Also confirmed that this works nicely together with https://github.com/godotengine/godot/pull/67055 to get exported arrays of custom node types.

### CODE REVIEW

**Comment by GuilhermeGSousa** [comment]
I'm closing this for now as this PR isn't really ready yet, there are still some issues on runtime. I'll reopen it once its ready

**Comment by GuilhermeGSousa** [comment]
Some known issues to fix before undrafting this:
- Exported arrays are not being saved correctly
- Exported typed arrays dont work for custom `Resources`, although that is outside of the scope of the issue this PR is trying to fix, and would probably deserve its own issue/PR

**Comment by GuilhermeGSousa** [comment]
@Paulb23 @YeldhamDev could I get a review for this one?

File: scene/resources/packed_scene.cpp [hunk 1]
```diff
@@ -600,6 +634,20 @@ Error SceneState::_parse_node(Node *p_owner, Node *p_node, int p_parent_idx, Has
...
 				value = missing_resource_properties[E.name];
 			}
+		} else if (E.type == Variant::ARRAY && E.hint == PROPERTY_HINT_TYPE_STRING) {
+			int hint_subtype_separator = E.hint_string.find(":");
+			if (hint_subtype_separator >= 0) {
+				String subtype_string = E.hint_string.substr(0, hint_subtype_separator);
+				int slash_pos = subtype_string.find("/");
+				PropertyHint subtype_hint = PropertyHint::PROPERTY_HINT_NONE;
+				if (slash_pos >= 0) {
+					subtype_hint = PropertyHint(subtype_string.substr(slash_pos + 1, subtype_string.size() - slash_pos - 1).to_int());
```
Comment (KoBeWi) [inline scene/resources/packed_scene.cpp:644]:
```suggestion
					subtype_hint = PropertyHint(subtype_string.get_slice("/", 1).to_int());
```
The whole code can be simplified using `get_slice()` or `get_slice_count()`.

**Comment by DmitriySalnikov** [comment]
I didn't know about this PR and tried to fix the problem myself, my code is below. Here I just cut off `metadata/_editor_prop_ptr_` if there is one (I don't know which option would be preferable 🤷‍♂️)
```diff
diff --git a/editor/editor_properties_array_dict.cpp b/editor/editor_properties_array_dict.cpp
index 451cb7cfee..606fc8239b 100644
--- a/editor/editor_properties_array_dict.cpp
+++ b/editor/editor_properties_array_dict.cpp
@@ -36,9 +36,10 @@
 #include "editor/editor_scale.h"
 #include "editor/editor_settings.h"
 #include "editor/inspector_dock.h"
+#include "scene/resources/packed_scene.h"
 
 bool EditorPropertyArrayObject::_set(const StringName &p_name, const Variant &p_value) {
-	String name = p_name;
+	String name = String(p_name).replace(SceneState::get_meta_pointer_property(""), "");
 
 	if (name.begins_with("indices")) {
 		int index = name.get_slicec('/', 1).to_int();
@@ -50,7 +51,7 @@ bool EditorPropertyArrayObject::_set(const StringName &p_name, const Variant &p_
 }
 
 bool EditorPropertyArrayObject::_get(const StringName &p_name, Variant &r_ret) const {
-	String name = p_name;
+	String name = String(p_name).replace(SceneState::get_meta_pointer_property(""), "");
 
 	if (name.begins_with("indices")) {
 		int index = name.get_slicec('/', 1).to_int();
@@ -159,8 +160,9 @@ EditorPropertyDictionaryObject::EditorPropertyDictionaryObject() {
 ///////////////////// ARRAY ///////////////////////////
 
 void EditorPropertyArray::_property_changed(const String &p_property, Variant p_value, const String &p_name, bool p_changing) {
-	if (p_property.begins_with("indices")) {
-		int index = p_property.get_slice("/", 1).to_int();
+	String name = p_property.replace(SceneState::get_meta_pointer_property(""), "");
+	if (name.begins_with("indices")) {
+		int index = name.get_slice("/", 1).to_int();
 		Variant array = object->get_array();
 		array.set(index, p_value);
 		emit_changed(get_edited_property(), array, "", true);

```

My changes and your changes for `PackedScene` work correctly (I couldn't apply all your changes via `git apply`). But only without `@tool` mode. With the `@tool` mode active, several errors appear when opening the scene and the array becomes `null`. Is this expected?

**Comment by lejar** [comment]
Hi, is there anything I can do to help move this along? I'd like to use this feature in a project.

**Comment by GuilhermeGSousa** [comment]
@lejar I just pushed @KoBeWi 's change request. I don't currently have my godot dev environment setup so I can't currently work on this PR, but everything should be here to fix the issue, its only missing a review.

**Comment by KoBeWi** [comment]
Needs rebase. Also this doesn't seem to work anymore, probably after TypedArray changes.
```
Attempted to set a variable of type 'NodePath' into a TypedArray of type 'Object'.
core\variant\array.cpp:410 - Condition "!_p->typed.validate(value, "set")" is true.
```

File: scene/resources/packed_scene.cpp [hunk 2]
```diff
@@ -404,7 +425,20 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
 
 	for (const DeferredNodePathProperties &dnp : deferred_node_paths) {
 		Node *other = dnp.base->get_node_or_null(dnp.path);
-		dnp.base->set(dnp.property, other);
+		if (String(dnp.property).contains("/indices/")) {
```
Comment (KoBeWi) [inline scene/resources/packed_scene.cpp:428]: Avoid converting `StringName` to `String` more than once. Do
```C++
String string_property = dnp.property;
```

File: scene/resources/packed_scene.cpp [hunk 3]
```diff
@@ -238,16 +238,37 @@ Node *SceneState::instantiate(GenEditState p_edit_state) const {
 
 					if (nprops[j].name & FLAG_PATH_PROPERTY_IS_NODE) {
 						uint32_t name_idx = nprops[j].name & (FLAG_PATH_PROPERTY_IS_NODE - 1);
+						NodeData::Property nprop = nprops[j];
+						Variant prop_variant = props[nprop.value];
+						StringName prop_name = snames[name_idx];
```
Comment (KoBeWi) [inline scene/resources/packed_scene.cpp:243]:
```suggestion
						const StringName &prop_name = snames[name_idx];
```
(technically applies to other variables too)

**Comment by GuilhermeGSousa** [comment]
Thanks @KoBeWi I'll check this out

**Comment by dannflor** [comment]
@GuilhermeGSousa are you still working on this? It's a pretty bad blocker for me and I'd like to see it merged sooner than later.

**Comment by GuilhermeGSousa** [comment]
Closing this in favor of https://github.com/godotengine/godot/pull/73256

**Pull request closed without merging**